pkg/dockerregistry/server/supermiddleware/registry.go

```diff
@@ -0,0 +1,78 @@
+package supermiddleware
+
+import (
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/reference"
+	"github.com/docker/distribution/registry/storage"
+	storagedriver "github.com/docker/distribution/registry/storage/driver"
+)
+
+type registry struct {
+	distribution.Namespace
+	inst *instance
+}
+
+func (reg *registry) Repository(ctx context.Context, named reference.Named) (distribution.Repository, error) {
+	repo, err := reg.Namespace.Repository(ctx, named)
+	if err != nil {
+		return repo, err
+	}
+	return reg.inst.Repository(ctx, repo, false)
+}
+
+// HackRegistry must not be used.
```

Comment:
so why is it here? can it be moved into a _test.go file?

Reply:
It's a huge hack and should be removed, it's here to make a private variable accessible from another module. We gave too much tasks to the type `*repository`. It implements distribution.Repository and it's responsible for getting data from the image stream. But, for example, pullthroughManifestService wants `*repository` and uses

  * m.repo.namespace
  * m.repo.name
  * m.repo.getImageOfImageStream
  * m.repo.rememberLayersOfManifest
  * m.repo.registryOSClient
  * m.repo.imageStreamGetter

In other words, it wants to fill cache and to get ImageStreamImage. But right now we can't get imageStreamGetter and cache without constructing entire *repository.

Splitting will be almost mechanical changes, but it will be pretty big and I don't want to mix it with this PR.